plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Utils.hs

```diff
@@ -311,38 +315,118 @@ costIsAcceptable = \case
+                    decSize $> True
+                | otherwise -> decSize $> True
+            Error{} -> decSize $> True
+            LamAbs _ _ _ body -> decSize >> go body
+            TyAbs _ _ _ body ->
+                -- `decSize` to account for the `TyAbs` node (which becomes `Delay` in UPLC).
+                decSize >> go body
+            Constr _ _ _ es ->
+                -- `decSize` twice to account for the `Constr` node and the index.
+                decSize >> decSize >> allM go es
+            Case _ _ arg cs -> decSize >> allM go (arg : cs)
+            IWrap _ _ _ body ->
+                -- No `decSize` here, because `IWrap` nodes disappear in UPLC.
+                go body
+            Unwrap _ body ->
+                -- No `decSize` here, because `Unwrap` nodes disappear in UPLC.
+                go body
+            -- Constants can be big! We could check the size here and inline if they're
+            -- small, but probably not worth it
+            Constant{} -> pure False
+            Apply _ fun arg ->
+                -- `decSize` to account for the `Apply` node.
+                decSize >> (go fun &&^ go arg)
+            TyInst _ body _ty ->
+                -- `decSize` to account for the `TyInst` node (which becomes `Force` in UPLC).
+                decSize >> go body
+            Let{} -> pure False
+
+    decSize = modify' (over stRemainingSize (\x -> x - 1))
 
 -- | Is this an utterly trivial type which might as well be inlined?
 trivialType :: Type tyname uni ann -> Bool
 trivialType = \case
     TyBuiltin{} -> True
     TyVar{}     -> True
     _           -> False
+
+{- Note [Size criteria for inlining]
+Inlining replaces LHS with RHS, for some `LHS = RHS`. The LHS is usually a variable and the
+RHS is usually the definition of the variable. But when we are considering inlining
+fully-saturated applications, LHS can have parameters.
+
+For example, given `x = \a b c d -> f y`, normally LHS is `x`. But when we are considering
+inlining a call site of `x` where `x` is fully applied, LHS becomes `x a b c d`, and
+RHS becomes `f y`.
+
+In order to avoid increasing program sizes (which is much more important for Plutus Core than
+it is for general-purpose programming languages), we should inline only if size(RHS) is
+no bigger than size(LHS).
+
+size(LHS) depends on the number of term parameters on the LHS. For `x = f y`, size(LHS) = 1.
+Each additional parameter increases size(LHS) by 2 (one for the `Apply` node, and one for
+the parameter itself). For `x a b c d = f y`, size(LHS) = 9.
+
+We ignore type parameters since types eventually get erased.
+
+To measure size(RHS), we have to be mindful of multiple occurrences of parameters.
+For example, in `x a b c = b a a`, we cannot consider size(RHS) to be no bigger than size(LHS)
+just because it is not bigger in the binding. Otherwise we will turn `x <large> b c` into
+`let a = <large> in b a a`, which can be worse than not inlining.
```

Comment:
Specifically because we won't get rid of the overhead of the binding/application? I had to think quite a lot to work out what this example was saying, worth elaborating.

So in general we _can't_ assume that a saturated application will reduce away entirely unless the arguments look good-to-inline. That actually seems like an attack on our whole argument for doing saturated inlining. So indeed your approach here seems critical. I think the note about saturated inlining needs updating, this is very important!